sympy/matrices/normalforms.py

```diff
@@ -0,0 +1,138 @@
+    >>> m = Matrix([[12, 6, 4], [3, 9, 6], [2, 16, 14]])
+    >>> setattr(m, "ring", ZZ)
+    >>> print(smith_normal_form(m))
+    Matrix([[1, 0, 0], [0, 10, 0], [0, 0, -30]])
+
+    '''
+    invs = smith_normal_invariants(m)
+    smf = diag(*invs)
+    n = len(invs)
+    if m.rows > n:
+        smf = smf.row_insert(m.rows, zeros(m.rows-n, m.cols))
+    elif m.cols > n:
+        smf = smf.col_insert(m.cols, zeros(m.rows, m.cols-n))
+    return smf
+
+def smith_normal_invariants(m, domain = None):
+    '''
+    Return the tuple of abelian invariants for a matrix `m`
+    (as in the Smith-Normal form)
+
+    '''
+    if not domain:
+        if not (hasattr(m, "ring") and m.ring.is_PID):
+            raise ValueError(
+                "The matrix entries must be over a principal ideal domain")
+        else:
+            domain = m.ring
+
+    m = m[:, :]
+
+    def add_rows(m, i, j, a, b, c, d):
+        # replace m[i, :] by a*m[i, :] + b*m[j, :]
+        # and m[j, :] by c*m[i, :] + d*m[j, :]
+        for k in range(m.cols):
+            e = m[i, k]
+            m[i, k] = a*e + b*m[j, k]
+            m[j, k] = c*e + d*m[j, k]
+
+    def add_columns(m, i, j, a, b, c, d):
+        # replace m[:, i] by a*m[:, i] + b*m[:, j]
+        # and m[:, j] by c*m[:, i] + d*m[:, j]
+        for k in range(m.rows):
+            e = m[k, i]
+            m[k, i] = a*e + b*m[k, j]
+            m[k, j] = c*e + d*m[k, j]
+
+    def clear_column(m):
+        # make m[1:, 0] zero by row and column operations
+        if m[0,0] == 0:
+            return m
+        pivot = m[0, 0]
+        for j in range(1, m.rows):
+            if m[j, 0] == 0:
+                continue
+            d, r = domain.div(m[j,0], pivot)
+            if r == 0:
+                add_rows(m, 0, j, 1, 0, -d, 1)
+            else:
+                a, b, g = domain.gcdex(pivot, m[j,0])
+                d_0 = domain.div(m[j, 0], g)[0]
+                d_j = domain.div(pivot, g)[0]
+                add_rows(m, 0, j, a, b, d_0, -d_j)
+                pivot = g
+        return m
+
+    def clear_row(m):
+        # make m[0, 1:] zero by row and column operations
+        if m[0] == 0:
+            return m
+        pivot = m[0, 0]
+        for j in range(1, m.cols):
+            if m[0, j] == 0:
+                continue
+            d, r = domain.div(m[0, j], pivot)
+            if r == 0:
+                add_columns(m, 0, j, 1, 0, -d, 1)
+            else:
+                a, b, g = domain.gcdex(pivot, m[0, j])
+                d_0 = domain.div(m[0, j], g)[0]
+                d_j = domain.div(pivot, g)[0]
+                add_columns(m, 0, j, a, b, d_0, -d_j)
+                pivot = g
+        return m
+
+    ind = [i for i in range(m.rows) if m[i,0] != 0]
+    if ind:
+        m = m.permute_rows([[0, ind[0]]])
+    else:
+        ind = [j for j in range(m.cols) if m[0,j] != 0]
+        if ind:
+            m = m.permute_cols([[0, ind[0]]])
+
```

Comment:
I think we should find the invariants in (divisibility) order starting with the gcd of all matrix entries and leaving the zeros, if any, to the end. Otherwise expressed, we should use the inverted inclusion order of the ideals generated by the invariants, first the largest one generated by all matrix entries, and the smallest ideals, `{0}` in particular, coming last.

This means that we should add 0 the the invariant list only when there are no nonzero entries left. Therefore all entries should be scanned at this point.